modelservice/src/main/java/com/expedia/adaptivealerting/modelservice/dao/es/ElasticSearchDetectorMappingService.java

```diff
@@ -0,0 +1,377 @@
+        Set<String> fields = getFields(createRequest);
+        Set<String> newFieldMappings = getFieldsWithoutExistingMapping(fields);
+        updateIndexMappings(newFieldMappings);
+        final DetectorMappingEntity detectorMappingEntity = new DetectorMappingEntity(createRequest.getUser(),
+            createRequest.getDetector(), QueryUtil.buildQuery(createRequest.getExpression()), true,
+            System.currentTimeMillis(), System.currentTimeMillis());
+        final IndexRequest indexRequest = new IndexRequest(elasticSearchConfig.getIndexName(),
+            elasticSearchConfig.getDocType());
+        indexRequest.source(DetectorMappingEntityToJson(detectorMappingEntity), XContentType.JSON);
+        try {
+            IndexResponse indexResponse = elasticSearchClient.index(indexRequest, RequestOptions.DEFAULT);
+            return indexResponse.getId();
+        } catch (IOException e) {
+            log.error(String.format("Indexing mapping %s failed", DetectorMappingEntityToJson(detectorMappingEntity), e));
+            throw new RuntimeException(e);
+        }
+    }
+    
+    private void updateIndexMappings(Set<String> newFieldMappings) {
+        PutMappingRequest request = new PutMappingRequest(elasticSearchConfig.getIndexName());
+        Map<String, Object> jsonMap = new HashMap<>();
+        Map<String, Object> type = new HashMap<>();
+        type.put("type", "keyword");
+        Map<String, Object> properties = new HashMap<>();
+        newFieldMappings.forEach(field -> {
+            properties.put(field, type);
+        });
+        jsonMap.put("properties", properties);
+        request.source(jsonMap);
+        request.type(elasticSearchConfig.getDocType());
+        try {
+            elasticSearchClient.indices().putMapping(request, RequestOptions.DEFAULT);
+        } catch (IOException e) {
+            log.error("Error updating mappings", e);
+            throw new RuntimeException(e);
+        }
+    }
+    
+    private Set<String> getFieldsWithoutExistingMapping(Set<String> fields) {
+        GetMappingsRequest request = new GetMappingsRequest();
+        request.indices(elasticSearchConfig.getIndexName());
+        
+        try {
+            GetMappingsResponse mappingsResponse = elasticSearchClient.indices().getMapping(request, RequestOptions.DEFAULT);
+            Set<String> mappeddFields = ((Map<String, String>)mappingsResponse.getMappings()
+                    .get(elasticSearchConfig.getIndexName())
+                    .get(elasticSearchConfig.getDocType()).sourceAsMap().get("properties")).entrySet().stream()
+                    .map(en -> en.getKey()).collect(Collectors.toSet());
+            fields.removeAll(mappeddFields);
+            return fields;
+        } catch (IOException e) {
+            log.error("Error finding mappings", e);
+            throw new RuntimeException(e);
+        }
+    }
+    
+    private Set<String> getFields(CreateDetectorMappingRequest createRequest) {
```

Comment:
This might make more sense as a public method on the `CreateDetectorMappingRequest` class, since it seems general-purpose enough and it's an operation on the contents of that class. Makes the code more coherent.

Reply:
Moved method.